pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +120,53 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	return pl.isReadWriteOncePodAccessModeConflict(pod)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as unschedulable and unresolvable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, ErrReasonNodeInfoError)
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			return framework.NewStatus(framework.Error, ErrReasonPVCError)
+		}
+
+		// If ReadWriteOncePod is used, it can be the only one on the PVC since it overrides any
+		// additional access modes. This is also because we do not know what access mode the pod
+		// intends to use if there are multiple.
+		if !(v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) && len(pvc.Spec.AccessModes) == 1) {
+			continue
+		}
+
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		for _, nodeInfo := range nodeInfos {
+			refCount, ok := nodeInfo.PVCRefCounts[key]
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if ok && utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && refCount > 0 {
+				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict)
```

Comment:
It's only `Unschedulable` not `UnschedulableAndUnresolvable`.

Unresolvable means that you can't schedule the pod even after preempting (evicting) other pods. In this case, you could preempt the other running pod that has the matching PVC.

Reply:
Even if that is technically possible, I am not sure if we want to allow this behavior. It seems like pod priorities would be a backdoor to usurping control over ReadWriteOncePod PVCs (and subsequently causing the previous pod to fail scheduling since the PVC is in use). I am new to the scheduler side of things so I may be off here, but is it common elsewhere to allow preemptions in cases where the previous workload will stop working?

@msau42 @jsafrane What are your thoughts on this?

Reply:
Priority is a limited resource, admins can setup quotas to limit who can use what priorities. https://kubernetes.io/docs/concepts/policy/resource-quotas/#limit-priority-class-consumption-by-default

Reply:
I see, thank you for the clarification